semconv/v1.7.0/http.go

```diff
@@ -293,3 +295,42 @@ func validateHTTPStatusCode(code int) (codes.Code, bool) {
 	}
 	return codes.Error, true
 }
+
+// SpanStatusFromHTTPStatusCode generates a status code and a message
+// as specified by the OpenTelemetry specification for a span.
+// Exclude 4xx for SERVER to set the appropriate status.
+func SpanStatusFromHTTPStatusCodeAndSpanKind(code int, spanKind trace.SpanKind) (codes.Code, string) {
+	spanCode, valid := validateHTTPStatusCodeAndSpanKind(code, spanKind)
+	if !valid {
+		return spanCode, fmt.Sprintf("Invalid HTTP status code %d", code)
+	}
+	return spanCode, ""
+}
+
+// Validates the HTTP status code and returns corresponding span status code.
+// If the `code` is not a valid HTTP status code, returns span status Error
+// and false. Exclude 4xx for SERVER to set the appropriate status.
+func validateHTTPStatusCodeAndSpanKind(code int, spanKind trace.SpanKind) (codes.Code, bool) {
+	category := code / 100
+	if spanKind == trace.SpanKindServer && category == 4 {
+		return codes.Unset, false
+	}
+	ranges, ok := validRangesPerCategory[category]
+	if !ok {
+		return codes.Error, false
+	}
+	ok = false
+	for _, crange := range ranges {
+		ok = crange.contains(code)
+		if ok {
+			break
+		}
+	}
+	if !ok {
+		return codes.Error, false
+	}
+	if category > 0 && category < 4 {
+		return codes.Unset, true
+	}
+	return codes.Error, true
```

Comment:
how about reusing `validateHTTPStatusCode` here?

Reply:
The purpose of adding a new function is to avoid break changes. If reusing `validateHTTPStatusCode` is better, I'll support and do that.

Reply:
I would prefer reusing `validateHTTPStatusCode` unless there is any strong reason not to do so. Copy-pasting can make bug fixing complex as it is harder to find "copy-pasted" code than a function call.

Reply:
ok, agree with u. I'll redo it by reusing `validateHTTPStatusCode` function.

Reply:
Sorry, but I think my comment was not clear enough. I meant something more or less like this:

go
```
// SpanStatusFromHTTPStatusCode generates a status code and a message
// as specified by the OpenTelemetry specification for a span.
// Exclude 4xx for SERVER to set the appropriate status.
func SpanStatusFromHTTPStatusCodeAndSpanKind(code int, spanKind trace.SpanKind) (codes.Code, string) {
	spanCode, valid := validateHTTPStatusCode(code)
	if !valid {
		return spanCode, fmt.Sprintf("Invalid HTTP status code %d", code)
	}
	category := code / 100
	if spanKind == trace.SpanKindServer && category == 4 {
		return codes.Unset, ""
	}
	return spanCode, ""
}

Reply:
I put the check after `validateHTTPSStatusCode` so that e.g. 412 HTTP Status Code will result in an error span

Reply:
Done.